beetle-core/src/main/java/com/xing/beetle/dedup/spi/KeyValueStoreBasedDeduplicator.java

```diff
@@ -90,7 +85,23 @@ public void deleteKeys(String messageId) {
         (beetleAmqpConfig.getBeetleRedisStatusKeyExpiryIntervalSeconds() > 0)
             ? Arrays.stream(keySuffixes).filter(s -> !s.equals(STATUS))
             : Arrays.stream(keySuffixes);
-    store.delete(suffixStream.map(s -> key(messageId, s)).toArray(String[]::new));
+    store.deleteMultiple(suffixStream.map(s -> key(messageId, s)).toArray(String[]::new));
+  }
+
+  /**
+   * returns true if the initialization succeeds, i.e. there were no already existing STATUS and
+   * EXPIRES keys. return false if at least one of STATUS and EXPIRES keys already exists.
+   */
+  @Override
+  public boolean initKeys(String messageId) {
+    Map<String, Value> keysValues = new HashMap<>();
+    keysValues.put(key(messageId, STATUS), new Value("incomplete"));
+    keysValues.put(
+        key(messageId, EXPIRES),
+        new Value(
+            System.currentTimeMillis()
+                + this.beetleAmqpConfig.getBeetleRedisStatusKeyExpiryIntervalSeconds() * 1000));
```

Comment:
Just to be clear: RedisStatusKeyExpiryIntervalSeconds needs to set as a TTL on the status key only after it has been set to "completed"(and in) if the app asks for tombstones by setting this to a non-zero value. The appropriate place would be the deleteKeys method. Because it needs to leave the key around when RedisStatusKeyExpiryIntervalSeconds is non-zero, but delete it if it's zero.

Reply:
BTW, this is another bug in the current implementation on master: setting a Redis key expiry on the status key before we have determined the status to be "complete" can result in Redis removing it before we're done processing all message instances.

Reply:
Yes, this PR fixes that as well: We do not set TTL for status key in the beginning anymore with this PR. 

In master, Status key is already left around when RedisStatusKeyExpiryIntervalSeconds is non-zero, but deleted it if it's zero as you can see [here in master](https://github.com/xing/java-beetle/blob/master/beetle-core/src/main/java/com/xing/beetle/dedup/spi/KeyValueStoreBasedDeduplicator.java#L90)